### ISSUE TITLE
Add support for simple structs?

### ISSUE THREAD

**Issue by sharkdp** [sharkdp/dbg-macro#6]
It would be cool to somehow add support for simple `struct`s/`class`es like
``` c++
struct CustomDatatype {
  int field1;
  std::string field2;
};

// …

CustomDatatype c{42, "hello"};
dbg(c);
```

We would somehow have to iterate over all members... without modifying to original struct definition. Not sure if this is possible in C++ (without reflection).

**Comment by hmenke** [comment]
You can use [magic_get](https://github.com/apolukhin/magic_get) (https://www.youtube.com/watch?v=abdeAew3gmQ).  Requires C++14.

**Comment by sharkdp** [comment]
`magic_get` looks fantastic, thank you for the reference.

If there is a small subset of the code that we could include in `dbg.h` (assuming there are no licensing problems, see similar discussion in #4 and #5), I think that would be a great addition. It looks like there is no way to get the actual name of the fields, so the best we could do would probably be an output like:
```
dbg(c); // CustomDatatype{42, "hello"}
```
(which would already be great).

**Comment by AlexBAV** [comment]
There's also [cista](https://github.com/felixguendling/cista), which uses another approach for simple type reflection (see [to_tuple.h](https://github.com/felixguendling/cista/blob/master/include/cista/reflection/to_tuple.h)) for example.

**Comment by sharkdp** [comment]
> There's also [cista](https://github.com/felixguendling/cista), which uses another approach for simple type reflection (see [to_tuple.h](https://github.com/felixguendling/cista/blob/master/include/cista/reflection/to_tuple.h) for example).

Thank you for the reference! Unfortunately, it seems to rely on C++17 functionality, but I would really like `dbg(…)` to be C++11 compatible.

**Comment by xorz57** [comment]
@sharkdp #31 Looks like we have a problem on Windows Build related with `Structs`.

**Comment by alexeyr** [comment]
For C++11 there is https://github.com/garbageslam/visit_struct but unfortunately it does need
```
VISITABLE_STRUCT(CustomDatatype, field1, field2);
```
Also, would it be good enough if `magic_get`-derived functionality was under `#ifdef __cplusplus`, so people using C++14 or later get it automatically and those limited to C++11 define `showValue` as usual?

Both are under B

**Comment by sharkdp** [comment]
Agreed. If the concerns above (license, code size) can be clarified, I'd be willing to accept this addition.

**Comment by nocoinman** [comment]
@sharkdp [Here](https://cppinsights.io/lnk?code=I2luY2x1ZGUgPGlvc3RyZWFtPgojaW5jbHVkZSA8c3RyaW5nPgojaW5jbHVkZSA8Y3N0ZGludD4KCm5hbWVzcGFjZSBkYmcgewogIHRlbXBsYXRlIDx0eXBlbmFtZSBUPgogIGlubGluZSBib29sIHByZXR0eV9wcmludChzdGQ6Om9zdHJlYW0mLCBjb25zdCBUJik7Cn0KCi8vIEJFR0lOOiBzdXBwb3J0IGZvciBzaW1wbGUgc3RydWN0cwoKLy8gQWNjZXB0IGFueSBudW1iZXIgb2YgYXJncyA+PSBOLCBidXQgZXhwYW5kIHRvIGp1c3QgdGhlIE50aCBvbmUuCi8vIEhlcmUsIE4gPT0gNi4KI2RlZmluZSBfR0VUX05USF9BUkcoXzEsIF8yLCBfMywgXzQsIF81LCBOLCAuLi4pIE4KI2RlZmluZSBDT1VOVF9WQVJBUkdTKC4uLikgX0dFVF9OVEhfQVJHKCJpZ25vcmVkIiwgIyNfX1ZBX0FSR1NfXywgNCwgMywgMiwgMSwgMCkKCi8vIERlZmluZSBzb21lIG1hY3JvcyB0byBoZWxwIHVzIGNyZWF0ZSBvdmVycmlkZXMgYmFzZWQgb24gdGhlCi8vIGFyaXR5IG9mIGEgZm9yLWVhY2gtc3R5bGUgbWFjcm8uCiNkZWZpbmUgX2ZlXzAoX2NhbGwsIC4uLikKI2RlZmluZSBfZmVfMShfY2FsbCwgeCkgX2NhbGwoeCkKI2RlZmluZSBfZmVfMihfY2FsbCwgeCwgLi4uKSBfY2FsbCh4KSBfZmVfMShfY2FsbCwgX19WQV9BUkdTX18pCiNkZWZpbmUgX2ZlXzMoX2NhbGwsIHgsIC4uLikgIF9jYWxsKHgpIF9mZV8yKF9jYWxsLCBfX1ZBX0FSR1NfXykKI2RlZmluZSBfZmVfNChfY2FsbCwgeCwgLi4uKSAgX2NhbGwoeCkgX2ZlXzMoX2NhbGwsIF9fVkFfQVJHU19fKQoKI2RlZmluZSBDQUxMX1hfRk9SX0VBQ0goeCwgLi4uKSBcCiAgICBfR0VUX05USF9BUkcoImlnbm9yZWQiLCAjI19fVkFfQVJHU19fLCBcCiAgICBfZmVfNCwgX2ZlXzMsIF9mZV8yLCBfZmVfMSwgX2ZlXzApKHgsICMjX19WQV9BUkdTX18pCgojZGVmaW5lIFBSSU5UX0ZJRUxEX1RPX1NUUkVBTSh4KSBcCiAgeyBcCiAgICBzdHJlYW0gPDwgIiciIDw8ICN4IDw8ICInOiAiOyBcCiAgICBwcmV0dHlfcHJpbnQoc3RyZWFtLCBpbnN0YW5jZS54KTsgXAogICAgaWYgKCsraV9maWVsZCA8IG5fZmllbGQpIHN0cmVhbSA8PCAiLCAiOyBcCiAgfQoKI2RlZmluZSBEQkdfUkVHSVNURVJfU1RSVUNUKEN1c3RvbURhdGFUeXBlLCAuLi4pIFwKICAgIG5hbWVzcGFjZSBkYmcgeyBcCiAgICAgIHRlbXBsYXRlPD4gXAogICAgICBpbmxpbmUgYm9vbCBwcmV0dHlfcHJpbnQoc3RkOjpvc3RyZWFtJiBzdHJlYW0sIGNvbnN0IEN1c3RvbURhdGFUeXBlJiBpbnN0YW5jZSkgeyBcCiAgICAgICAgc2l6ZV90IGlfZmllbGQgPSAwLCBuX2ZpZWxkID0gQ09VTlRfVkFSQVJHUyhfX1ZBX0FSR1NfXyk7IFwKICAgICAgICBzdHJlYW0gPDwgInsiOyBcCiAgICAgICAgICBDQUxMX1hfRk9SX0VBQ0goUFJJTlRfRklFTERfVE9fU1RSRUFNLCBfX1ZBX0FSR1NfXykgXAogICAgICAgIHN0cmVhbSA8PCAifSI7IFwKICAgIHJldHVybiB0cnVlOyBcCiAgICAgIH0gXAogIH0gCgovLyBFTkQgIDogc3VwcG9ydCBmb3Igc2ltcGxlIHN0cnVjdHMKCnN0cnVjdCBNZXRhRGF0YSB7CiAgICBzdGQ6OnN0cmluZyBuYW1lOwogICAgaW50NjRfdCBpZDsKfTsKCnN0cnVjdCBFbnRpdHkzRCB7CiAgICBpbnQzMl90IHgsIHksIHo7CiAgICBNZXRhRGF0YSBtZXRhOwp9OwoKREJHX1JFR0lTVEVSX1NUUlVDVChNZXRhRGF0YSwgbmFtZSwgaWQpCkRCR19SRUdJU1RFUl9TVFJVQ1QoRW50aXR5M0QsIHgsIHksIHosIG1ldGEpCg==&insightsOptions=cpp17&std=cpp17&rev=1.0) is a proposal to incorporate this feature. This is a working example that I came up with based on the above discussions and some snippets from [here](https://codecraft.co/2014/11/25/variadic-macros-tricks/).

Sample driver code
---

```cpp
struct MetaData {
    std::string name;
    int64_t id;
};

struct Entity3D {
    int32_t x, y, z;
    MetaData meta;
};

DBG_REGISTER_STRUCT(MetaData, name, id)
DBG_REGISTER_STRUCT(Entity3D, x, y, z, meta)

int32_t main() {
    Entity3D e = {3, 2, 1, {"dummy_name", 98765}};
    dbg(e);  // [main.cc:34 (main)] e = {'x': 3, 'y': 2, 'z': 1, 'meta': {'name': "dummy_name", 'id': 98765}} (Entity3D)
}
```

If this is satisfactory I'll be glad to open a PR!

**Comment by sharkdp** [comment]
Thank you. If we add support for this, then I would definitely look for a solution which does not require any additional macro calls like `DBG_REGISTER_STRUCT`.

**Comment by Smalldy** [comment]
> Thank you. If we add support for this, then I would definitely look for a solution which does not require any additional macro calls like `DBG_REGISTER_STRUCT`.


This is too idealistic, almost impossible I think.
